torch/csrc/jit/codegen/cuda/lower_utils.cpp

```diff
@@ -208,6 +208,66 @@ kir::Expr* applyReplacements(
   }
 }
 
+c10::optional<IterDomain*> getMaybeWarpReductionDim(
+    const kir::ReductionOp* node) {
+  kir::TensorView* kir_tv = nullptr;
+  auto kir_ti = dynamic_cast<kir::TensorIndex*>(node->out());
+  if (kir_ti) {
+    kir_tv = kir_ti->view();
+  } else {
+    kir_tv = dynamic_cast<kir::TensorView*>(node->out());
+  }
+  if (!kir_tv) {
+    return c10::nullopt;
+  }
+  auto fuser_reduction = kir_tv->fuserTv()->definition()->as<ReductionOp>();
+  return getMaybeWarpReductionDim(fuser_reduction);
+}
+
+c10::optional<IterDomain*> getMaybeWarpReductionDim(const ReductionOp* node) {
+  auto fuser_tv_out = node->out()->as<TensorView>();
+  auto fuser_tv_in = node->in()->as<TensorView>();
+
+  // only support reducing to registers for now.
+  if (fuser_tv_in->getMemoryType() != MemoryType::Local ||
+      fuser_tv_out->getMemoryType() != MemoryType::Local) {
+    return c10::nullopt;
+  }
+
+  IterDomain* reduction_on_xdim = nullptr;
+  for (auto id : fuser_tv_out->domain()->domain()) {
+    // Currently warp reduction only allows
+    //  serial and block.x parallel reductions
+    if (id->isReduction() && id->isParallelized()) {
+      if (id->getParallelType() == ParallelType::TIDx) {
+        reduction_on_xdim = id;
+      } else {
+        return c10::nullopt;
+      }
+    }
+  }
+  if (!reduction_on_xdim) {
+    return c10::nullopt;
+  }
+
+  if (!reduction_on_xdim->start()->isZeroInt()) {
+    return c10::nullopt;
+  }
```

Comment:
Is this a hard constraint? I'm adding support of reduction domains with non-zero start at #1051 and #1026. It seems that it should work fine with a similar extension as in #1051.

Reply:
Great question! I just disabled it because I was unsure yet about the whole picture about using non-zero start and what cases warp reduce needs to accommodate. Could you share with me a few test names so I could try to extend and remove this check?

Reply:
It would need non-padded shift in #1026, which still has issues to address. I think the warp reduce should work as long as it separates the read and write predicates as in #1051. You could just leave it as is and file an issue so that I could take a look once #1026 is completed.

Reply:
Sure I will rebase and add support for separate predicates.

Reply:
Issue created and to be addressed in a follow up #1064